core/ipc/rpc/kafka/src/main/java/org/opennms/core/ipc/rpc/kafka/KafkaRpcClientFactory.java

```diff
@@ -0,0 +1,311 @@
+                Logging.putPrefix(RpcClientFactory.LOG_PREFIX);
+                if (request.getLocation() == null || request.getLocation().equals(location)) {
+                    // The request is for the current location, invoke it directly
+                    return module.execute(request);
+                }
+                final JmsQueueNameFactory topicNameFactory = new JmsQueueNameFactory("rpc-request", module.getId(),
+                        request.getLocation());
+                String requestTopic = topicNameFactory.getName();
+                String marshalledRequest = module.marshalRequest(request);
+                // Generate RPC Id for every request to track request/response.
+                String rpcId = UUID.randomUUID().toString();
+                // Calculate timeout based on ttl and default timeouts.
+                long metadataRefreshRate = Long.parseLong(TOPICS_METADATA_MAX_AGE);
+                long ttl = request.getTimeToLiveMs() != null ? request.getTimeToLiveMs().longValue() : 0;
+                // timeout for RPC > metadataRefreshRate, add default timeout to metadata refresh rate.
+                long timeout = metadataRefreshRate + Long.parseLong(TIMEOUT_FOR_KAFKA_RPC);
+                // ttl should be > default timeout otherwise use default.
+                timeout = ttl < timeout ? timeout : ttl;
+                // Create a future and add it to response handler which will complete the future when it receives callback.
+                final CompletableFuture<T> future = new CompletableFuture<>();
+                ResponseHandler<S, T> responseHandler = new ResponseHandler<S, T>(future, module, rpcId,
+                        System.currentTimeMillis() + timeout);
+                delayQueue.offer(responseHandler);
+                rpcResponseMap.put(rpcId, responseHandler);
+                if (request.getSystemId() != null) {
+                    // For directed RPCs, send request to all partitions (consumers),
+                    // this forces partitions >= max(number of minions at location)
+                    RpcMessageProtos.RpcMessage rpcMessage = RpcMessageProtos.RpcMessage.newBuilder()
+                                                                 .setRpcId(rpcId)
+                                                                 .setSystemId(request.getSystemId())
+                                                                 .setRpcContent(ByteString.copyFromUtf8(marshalledRequest))
+                                                                 .build();
+                    List<PartitionInfo> partitionInfo = producer.partitionsFor(requestTopic);
+                    partitionInfo.stream().forEach(partition -> {
+                        // Use systemId as key.
+                        final ProducerRecord<String, byte[]> record = new ProducerRecord<>(requestTopic,
+                                partition.partition(), request.getSystemId(), rpcMessage.toByteArray());
```

Comment:
Is there are specific reason we use the systemId as the key here? Do we just need *some* key in order to be able to specify the partition, or does it have some special meaning?

Reply:
yeah, ProducerRecord needs some "key" when we specify partition

Reply:
Thinking about this some more... I think we should use the message UUID as the partition key for all messages in both the request and response topic. This should have a uniform distribution and will allow external consumers to easily join both streams to match requests and responses.